File: Seccion_IV_Polynomial Regression/polinomial_regression.R
# Regresión polinómica

#Importamos el dataset
dataset = read.csv("Position_Salaries.csv")
dataset =dataset[, 2:3]

#Division de los datos para la fase de entrenamiento y para la fase de testing
# install.packages("caTools")
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)

# #Escalado de valores numericos
# training_set[,2:3] = scale(testing_set[,2:3])
# testing_set[,2:3] = scale(testing_set[,2:3])

# Ajuste del modelo de regresión lineal con el conjunto de datos
lin_reg = lm(formula = Salary ~ ., 
             data = dataset)

# Ajuste del modelo de regresión polinómica con el conjunto de datos
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

poly_reg = lm(formula = Salary ~ .,
              data = dataset)

# Predicción de nuevos resultados con Regresión Lineal
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))


# Predicción de nuevos resultados con Regresión Polinómica
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))

# Visualización del modelo lineal
# install.packages("ggplot2")
# library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("Predicción lineal del sueldo en función del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")


# Visualización del modelo polinómico
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x = x_grid, y = predict(poly_reg, 
                                        newdata = data.frame(Level = x_grid,
                                                             Level2 = x_grid^2,
                                                             Level3 = x_grid^3,
                                                             Level4 = x_grid^4))),
            color = "blue") +
  ggtitle("Predicción polinómica del suedlo en función del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

File: Seccion_IV_Polynomial Regression/polinomial_regression.py
#Regresión polinómica

#Importamos librerias 
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd 

#Importamos el dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:,1:2].values
y = dataset.iloc[:,-1].values

"""
#Dividimos el dataset en conj de datos para el entrenamiento y datos para testing 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)


#Escalado de variables 
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""

#Ajuste del modelo de regresion lineal simple

from sklearn.linear_model import LinearRegression
regression_ls = LinearRegression()
regression_ls.fit(X, y)

#Ajuste de la regresion polinomica con el dataset 

from sklearn.preprocessing import PolynomialFeatures
regression_lp = PolynomialFeatures(degree = 4)
X_poly = regression_lp.fit_transform(X)

linear_reg2 = LinearRegression()
linear_reg2.fit(X_poly, y)

#Visualizacion de los resultados del modelo lineal
plt.scatter(X, y, color = "red")
plt.plot(X, regression_ls.predict(X), color = "blue")
plt.title("Modelo de Regresión Lineal")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

#Visualizacion de los resultados del modelo polinómico

X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = "red")
plt.plot(X_grid, linear_reg2.predict(regression_lp.fit_transform(X_grid)), color = "blue")
plt.title("Modelo de Regresión Polinómica")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# Predicción de los modelos
# Sintaxis de doble corchete para hacer la predicción en las últimas versiones de Python (3.7+)
regression_ls.predict([[6.5]])
linear_reg2.predict(regression_lp.fit_transform([[6.5]]))














